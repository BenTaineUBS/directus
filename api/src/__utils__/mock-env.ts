import { afterEach, vi } from 'vitest';

let env = {};

export function setEnv(newEnv: Record<string, string>) {
	env = { ...env, ...newEnv };
}

/** Static env mock - to be used inside `vi.mock` */
export async function mockEnv(options?: { env?: Record<string, string>; withDefaults?: boolean }) {
	const initialEnv = options?.env ?? {};
	const withDefaults = options && 'withDefaults' in options ? options.withDefaults : true;

	env = { ...initialEnv };

	afterEach(() => {
		env = { ...initialEnv };
	});

	const { defaults, processValues } = await vi.importActual<typeof import('../env.js')>('../env.js');

	return {
		get default() {
			return processValues({ ...(withDefaults && defaults), ...env });
		},
	};
}

/** Dynamic env mock */
export function doMockEnv(options?: { env?: Record<string, string>; withDefaults?: boolean }) {
	const initialEnv = options?.env ?? {};
	const withDefaults = options && 'withDefaults' in options ? options.withDefaults : true;

	let env = { ...initialEnv };

	vi.doMock('../env.js', async () => {
		const { defaults, processValues } = await vi.importActual<typeof import('../env.js')>('../env.js');

		return {
			get default() {
				return processValues({ ...(withDefaults && defaults), ...env });
			},
		};
	});

	afterEach(() => {
		env = { ...initialEnv };
	});

	return { env, setEnv, replaceEnv };

	function setEnv(newEnv: Record<string, string>) {
		env = { ...env, ...newEnv };
	}

	function replaceEnv(newEnv: Record<string, string>) {
		env = { ...newEnv };
	}
}
