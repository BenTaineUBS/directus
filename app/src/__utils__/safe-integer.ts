type numeric = number | bigint;

export class SafeInteger {
	private _value: numeric;
	private MAX_VALUE: numeric;
	private MIN_VALUE: numeric;

	constructor(value: number | bigint | string, isBigInt = false) {
		if (isBigInt) {
			this._value = BigInt(value);
			this.MAX_VALUE = BigInt(9223372036854775807n);
			this.MIN_VALUE = BigInt(-9223372036854775807n);
			return;
		}

		this._value = Number(value);
		this.MAX_VALUE = Number.MAX_SAFE_INTEGER;
		this.MIN_VALUE = Number.MIN_SAFE_INTEGER;
	}

	get isMaximum() {
		return this._value >= this.MAX_VALUE;
	}

	get isMinimum() {
		return this._value <= this.MIN_VALUE;
	}

	get isValid() {
		return !this.isMaximum && !this.isMinimum;
	}

	get isInvalid() {
		return !this.isValid;
	}

	get value() {
		return typeof this._value === 'bigint' ? this._value.toString() : this._value;
	}

	toString() {
		return this._value.toString();
	}

	add(value: number) {
		if (this.isMaximum) {
			return;
		}

		if (typeof this._value === 'bigint') {
			this._value += BigInt(value);
			return;
		}

		this._value += value;
	}

	sub(value: number) {
		if (this.isMinimum) {
			return;
		}

		if (typeof this._value === 'bigint') {
			this._value -= BigInt(value);
			return;
		}

		this._value -= value;
	}

	increment() {
		this.add(1);
	}

	decrement() {
		this.sub(1);
	}
}
