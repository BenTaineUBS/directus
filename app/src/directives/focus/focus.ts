import { Directive } from 'vue';

const Focus: Directive = {
	inserted(el, binding) {
		if (binding.value) {
			el.focus();
		} else {
			el.blur();
		}
	},
};

export default Focus;
