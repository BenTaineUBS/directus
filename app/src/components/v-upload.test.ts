import type { GlobalMountOptions } from '@/__utils__/types';
import { Focus } from '@/__utils__/focus';
import { mount } from '@vue/test-utils';
import { expect, test, beforeEach, vi, describe } from 'vitest';
import { createI18n } from 'vue-i18n';
import { Tooltip } from '../__utils__/tooltip';
import { createTestingPinia } from '@pinia/testing';
import { setActivePinia } from 'pinia';
import VButton from './v-button.vue';
import VUpload from './v-upload.vue';

vi.mock('vue-router', () => ({
	useRoute: vi.fn(),
	useLink: vi.fn().mockImplementation(() => ({
		route: vi.fn().mockReturnValue(''),
		isActive: vi.fn().mockReturnValue(false),
		isExactActive: vi.fn().mockReturnValue(false),
	})),
}));

const i18n = createI18n({
	legacy: false,
	messages: {
		'en-US': {
			click_to_browse: 'Click to Browse',
			drag_file_here: 'Drag & Drop a File Here',
			import_from_url: 'Import File from URL',
			folders: 'Folders',
		},
	},
});

const global: GlobalMountOptions = {
	stubs: [
		'v-icon',
		'v-progress-linear',
		'v-card-title',
		'v-input',
		'v-card-text',
		'v-card-actions',
		'v-card',
		'v-dialog',
		'v-progress-circular',
	],
	plugins: [i18n],
	directives: {
		Tooltip,
		Focus,
	},
	components: {
		'v-button': VButton,
	},
};

Object.defineProperty(window, 'FormData', {
	writable: true,
	value: FormData,
});

vi.mock('@/utils/notify', () => ({
	notify: vi.fn(),
}));

vi.mock('@/api', () => {
	return {
		default: {
			post: vi.fn(),
			patch: vi.fn(),
		},
	};
});

describe('V-Upload', () => {
	beforeEach(() => {
		setActivePinia(
			createTestingPinia({
				createSpy: vi.fn,
				stubActions: false,
			}),
		);
	});

	const props = {
		multiple: false,
		preset: undefined,
		fileId: undefined,
		fromUser: true,
		fromUrl: false,
		fromLibrary: false,
	};

	test('Mount component', () => {
		expect(VUpload).toBeTruthy();

		const wrapper = mount(VUpload, {
			props,
			global,
		});

		expect(wrapper.html()).toMatchSnapshot();
	});

	test('Should upload new file', async () => {
		const wrapper = mount(VUpload, {
			props,
			global,
		});

		const file = new File(['image'], 'image.png', { type: 'image/png' });

		const fileUpload = wrapper.get<HTMLInputElement>('input[type="file"]');

		expect(fileUpload).toBeTruthy();

		Object.defineProperty(fileUpload.element, 'files', {
			value: [file],
			writable: false,
		});

		await fileUpload.trigger('change');
		await wrapper.vm.$nextTick();

		expect(fileUpload.element.files).toEqual([file]);
	});
});
