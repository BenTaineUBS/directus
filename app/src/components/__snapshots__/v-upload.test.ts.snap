// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`V-Upload > Mount component 1`] = `
"<div data-v-88c1425a="" data-dropzone="" class="v-upload">
  <div data-v-88c1425a="" class="actions">
    <div data-v-28d526fe="" data-v-88c1425a="" class="v-button secondary rounded"><button data-v-28d526fe="" class="button align-center icon normal" type="button"><span data-v-28d526fe="" class="content"><input data-v-88c1425a="" class="browse" type="file"><v-icon-stub data-v-88c1425a="" name="file_upload"></v-icon-stub></span>
        <div data-v-28d526fe="" class="spinner">
          <!--v-if-->
        </div>
      </button></div>
    <!--v-if-->
    <!--v-if-->
  </div>
  <p data-v-88c1425a="" class="type-label">Drag & Drop a File Here</p>
  <!--v-if-->
</div>"
`;
